transforms/universal/fdedup/python/src/cluster_analysis_transform.py

```diff
@@ -0,0 +1,336 @@
+jaccard_similarity_threshold_default = 0.75
+""" Default Jaccard similarity threshold (from FineWeb https://arxiv.org/pdf/2406.17557)"""
+num_segments_default = 1
+""" Default number of segments dividing the hashing space for each band"""
+sort_output_default = False
+
+
+class ClusterAnalysisTransform(AbstractFolderTransform):
+    """
+    This is the second transform of the fuzzy dedup pipeline. It runs in parallel:
+    for each band, the hashing interval is divided into segments. A cluster analysis
+    uses as input all the parquet files from segment of a band. The `bands` output
+    of the signature calculation, the first transform in the fuzzy dedup pipeline
+    contains all the data for a given segment s of a specific band b in the
+    subfolder `bands/band=b/segment=s`.
+    The transform loads all the parquet files in the `bands/band=b/segment=s`
+    subfolder. Each one of these parquet files has two columns: the `band_hash`
+    and a `data` structure, which includes the `document_id`, the `minhashes` and
+    the `document_size` fields. Once all the files have been loaded in a single
+    dataframe, a `group_by` operation on the `band_hash` field is performed in
+    that dataframe. All the documents that have the same band_hash are grouped
+    in a cluster. Subsequently, the documents of each cluster are sorted in
+    descending order according to their size, and a Jaccard similarity is
+    calculated between the cluster documents. The documents for which the Jaccard
+    similarity is above the `jaccard_similarity_threshold` remain in the cluster,
+    the others are removed from the cluster. Finally, from each cluster that has
+    more than one document after running the Jaccard similarity, we select a doc
+    to keep (the largest size document), and mark the other documents as
+    duplicates. The resulting clusters are saved in a file for further analysis.
+
+    Args:
+        num_bands: number of bands used in the banding technique
+        jaccard_similarity_threshold: Jaccard similarity threshold above which two documents are duplicates
+        num_segments: the number of segments dividing the hashing space for each band
+    """
+
+    def __init__(self, config: dict[str, Any]):
+        """
+        Initialize based on the dictionary of configuration information.
+        This is generally called with configuration parsed from the CLI arguments
+        defined by the companion runtime, ClusterAnalysisTransformRuntime.
+        """
+        super().__init__(config)
+        self.num_bands = config.get(num_bands_key, num_bands_default)
+        self.num_segments = config.get(num_segments_key, num_segments_default)
+        self.jaccard_similarity_threshold = config.get(
+            jaccard_similarity_threshold_key, jaccard_similarity_threshold_default
+        )
+        self.sort_output = config.get(sort_output_key, sort_output_default)
+        self.data_access = config.get("data_access")
```

Comment:
If data_access is not provided, what happens.  Either throw an exception, or use DataAccessLocal() as the default?

Reply:
Fixed in commit https://github.com/IBM/data-prep-kit/pull/699/commits/8de14b51e55a69bfa7f4dddc763500fb3a01db4d